server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotsServiceIT.java

```diff
@@ -0,0 +1,100 @@
+        createIndexWithRandomDocs("test-index", randomIntBetween(1, 42));
+        createSnapshot("test-repo", "test-snapshot", List.of("test-index"));
+
+        final MockLogAppender mockLogAppender = new MockLogAppender();
+
+        try {
+            mockLogAppender.start();
+            Loggers.addAppender(LogManager.getLogger(SnapshotsService.class), mockLogAppender);
+
+            mockLogAppender.addExpectation(
+                new MockLogAppender.UnseenEventExpectation(
+                    "[does-not-exist]",
+                    SnapshotsService.class.getName(),
+                    Level.INFO,
+                    "deleting snapshots [does-not-exist] from repository [test-repo]"
+                )
+            );
+
+            mockLogAppender.addExpectation(
+                new MockLogAppender.SeenEventExpectation(
+                    "[test-snapshot]",
+                    SnapshotsService.class.getName(),
+                    Level.INFO,
+                    "deleting snapshots [test-snapshot] from repository [test-repo]"
+                )
+            );
+
+            final SnapshotMissingException e = expectThrows(
+                SnapshotMissingException.class,
+                () -> startDeleteSnapshot("test-repo", "does-not-exist").actionGet()
+            );
+            assertThat(e.getMessage(), containsString("[test-repo:does-not-exist] is missing"));
+            assertThat(startDeleteSnapshot("test-repo", "test-snapshot").actionGet().isAcknowledged(), is(true));
+
+            mockLogAppender.assertAllExpectationsMatched();
+        } finally {
+            Loggers.removeAppender(LogManager.getLogger(SnapshotsService.class), mockLogAppender);
+            mockLogAppender.stop();
+            deleteRepository("test-repo");
+        }
+    }
+
+    public void testSnapshotDeletionFailureShouldBeLogged() throws InterruptedException {
+        createRepository("test-repo", "mock");
+        createIndexWithRandomDocs("test-index", randomIntBetween(1, 42));
+        createSnapshot("test-repo", "test-snapshot", List.of("test-index"));
+
+        final MockLogAppender mockLogAppender = new MockLogAppender();
+
+        try {
+            mockLogAppender.start();
+            Loggers.addAppender(LogManager.getLogger(SnapshotsService.class), mockLogAppender);
+
+            mockLogAppender.addExpectation(
+                new MockLogAppender.SeenEventExpectation(
+                    "[test-snapshot]",
+                    SnapshotsService.class.getName(),
+                    Level.WARN,
+                    "failed to complete snapshot deletion for [test-snapshot] from repository [test-repo]"
+                )
+            );
+
+            final MockRepository mockRepository = getRepositoryOnMaster("test-repo");
+            mockRepository.setRandomControlIOExceptionRate(1.0);
+            final Exception e = expectThrows(Exception.class, () -> startDeleteSnapshot("test-repo", "test-snapshot").actionGet());
```

Comment:
Could we use `setBlockAndFailOnWriteIndexFile()` so that the deletion starts successfully and then fails during execution? Possibly randomly choose between the two failure mechanisms.

Reply:
Good call. updated